### DISCUSSION TITLE
Advanced Anti Cheat (AAC) Compatibility Hook

### DISCUSSION THREAD

**Discussion by hamcoremusic** [Zrips/CMI#5007] [Feature Requests]
Hello!

As a server owner, of course I want to block all the hackers that try and ruin the game. You could just sit there and monitor every players movement or playstyle individually... or you can get an anti-cheat plugin to help you with that.

Now I'm not here to sell you on anti-cheat software, but rather ask for some CMI compatibility for one of the most used anti-cheat plugins available!

The developer of the plugin has some API you can implement into CMI that will help prevent false positives. If you look below, I have provided some links/resources that I think could help with CMI getting detected for some cheats (some of my players were getting detected for teleporting, and others were getting dinged for console run `cmi launch` commands).

[Resource link](https://www.spigotmc.org/resources/aac-advanced-anti-cheat-hack-kill-aura-blocker.6442/)

> Many plugins can be made compatible with AAC with relatively little effort:
>- Make sure any Bukkit events that you fire are subclasses of the actual event (e.g. fire FakeBlockBreakEvent instead of BlockBreakEvent)
>- Make sure you don't modify or send packets on behalf of a player
>- As a last resort, hook into AAC's API and call addExemption and removeExemption as necessary.

The API is listed in the resource page under a spoiler button. They also provide [API-only jars](https://github.com/konsolas/AAC-Issues) if that's a better way to look at it. I would stick with AAC v5.0.0 compatibility.

I think this would be a great addition and another decent selling point for CMI.

**Comment by hamcoremusic** [comment]
Another thing that may be worth noting is that AAC uses ProtocolLib to track specific events. This is what it is listening to:
```
PacketAdapter[plugin=AAC5, sending=ListeningWhitelist[priority=LOW, packets=[MAP_CHUNK[class=PacketPlayOutMapChunk, id=32], ARM_ANIMATION[class=PacketPlayInArmAnimation, id=44], ENTITY_VELOCITY[class=PacketPlayOutEntityVelocity, id=70], TELEPORT_ACCEPT[class=PacketPlayInTeleportAccept, id=0], MULTI_BLOCK_CHANGE[class=PacketPlayOutMultiBlockChange, id=59], ENTITY_ACTION[class=PacketPlayInEntityAction, id=28], ABILITIES[class=PacketPlayInAbilities, id=26], BLOCK_DIG[class=PacketPlayInBlockDig, id=27], POSITION[class=PacketPlayOutPosition, id=52], VEHICLE_MOVE[class=PacketPlayInVehicleMove, id=22], LOOK[class=PacketPlayInLook, id=20], FLYING[class=PacketPlayInFlying, id=21], POSITION[class=PacketPlayInPosition, id=18], POSITION_LOOK[class=PacketPlayInPositionLook, id=19], BLOCK_CHANGE[class=PacketPlayOutBlockChange, id=11], USE_ENTITY[class=PacketPlayInUseEntity, id=14], USE_ITEM[class=PacketPlayInUseItem, id=46], BLOCK_PLACE[class=PacketPlayInBlockPlace, id=47]], gamephase=PLAYING, options=[ASYNC]], receiving=ListeningWhitelist[priority=LOW, packets=[MAP_CHUNK[class=PacketPlayOutMapChunk, id=32], ARM_ANIMATION[class=PacketPlayInArmAnimation, id=44], ENTITY_VELOCITY[class=PacketPlayOutEntityVelocity, id=70], TELEPORT_ACCEPT[class=PacketPlayInTeleportAccept, id=0], MULTI_BLOCK_CHANGE[class=PacketPlayOutMultiBlockChange, id=59], ENTITY_ACTION[class=PacketPlayInEntityAction, id=28], ABILITIES[class=PacketPlayInAbilities, id=26], BLOCK_DIG[class=PacketPlayInBlockDig, id=27], POSITION[class=PacketPlayOutPosition, id=52], VEHICLE_MOVE[class=PacketPlayInVehicleMove, id=22], LOOK[class=PacketPlayInLook, id=20], FLYING[class=PacketPlayInFlying, id=21], POSITION[class=PacketPlayInPosition, id=18], POSITION_LOOK[class=PacketPlayInPositionLook, id=19], BLOCK_CHANGE[class=PacketPlayOutBlockChange, id=11], USE_ENTITY[class=PacketPlayInUseEntity, id=14], USE_ITEM[class=PacketPlayInUseItem, id=46], BLOCK_PLACE[class=PacketPlayInBlockPlace, id=47]], gamephase=PLAYING, options=[ASYNC]]]
```

I don't know if that's useful or not, but this is on the latest ProtocolLib dev build on 1.16.5

**Comment by ChiboYen** [comment]
+1 for this suggestion, if CMI is to be considered a plugin worthy of usage on large servers, then it must have compatibility with other plugins of the same caliber. Please consider implementing this Zrips.

**Comment by Parthurax** [comment]
+1

**Comment by Ace0fspadev** [comment]
+1

**Comment by AZFire1** [comment]
+1

**Comment by EuanM199** [comment]
+1

**Comment by hamcoremusic** [comment]
Could we get some kind of response on this? I would love to see what the dev thinks :)

**Comment by 4ubuntu** [comment]
+1

**Comment by TomLewis** [comment]
Same but for matrix.

**Comment by hamcoremusic** [comment]
Bump. It's been a while and I would love to know what the developer thinks about this!
